File: LectureCodes/Evp/gradit/MATLAB/gradit.m
function x = gradit(A,b,x,rtol,atol,maxit)
r = b-A*x; % residual $\to$ Def.~\ref{def:residual}
for k=1:maxit
  p = A*r;
  ts = (r'*r)/(r'*p); % \emph{cf.} \eqref{eq:tmin}
  x = x + ts*r;
  cn = (abs(ts)*norm(r); % norm of correction
  if ((cn < rtol*norm(x)) || (cn < atol))
    return;
  end
  r = r - ts*p; % \label{gi:10}
end

File: LectureCodes/Evp/gradit/Eigen/gradit.cpp
///////////////////////////////////////////////////////////////////////////
/// Demonstration code for lecture "Numerical Methods for CSE" @ ETH Zurich
/// Porting gradit.m to C++/Eigen.
/// (C) 2020 SAM, D-MATH
/// Author(s): William Andersson
/// Repository: https://gitlab.math.ethz.ch/NumCSE/NumCSE/
/// Do not remove this header.
//////////////////////////////////////////////////////////////////////////

#include "gradit.hpp"

VectorXd gradit(MatrixXd A, VectorXd b, VectorXd x, double rtol, double atol, unsigned int maxit) {
    // Residual
    VectorXd r = b - A * x;

    for (int k = 0; k < maxit; k++) {
        VectorXd p = A * r;

        double numerator = (r.transpose() * r);
        double denominator = (r.transpose() * p);
        double ts = numerator / denominator;  // cf. tmin

        x += ts * r;

        double cn = std::abs(ts) * r.norm();  // norm of correction

        if (cn < rtol * x.norm() || cn < atol) {
            break;
        }

        r -= ts * p;
    }
